### ISSUE TITLE
suggest `await` for future-related type errors

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#61076]
A common problem when using async-await is to forget to invoke `.await`. This often shows up as a type error, e.g. in a case like this ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ef1caa955ef1aceadf33aadd348e646c)):

```rust
#![feature(async_await)]

async fn make_u32() -> u32 {
    22
}

fn take_u32(x: u32) {

}

async fn foo() {
    let x = make_u32();
    take_u32(x);
}

fn main() { }
```

currently we give a rather generic error:

```
error[E0308]: mismatched types
  --> src/main.rs:13:14
   |
13 |     take_u32(x);
   |              ^ expected u32, found opaque type
   |
   = note: expected type `u32`
              found type `impl std::future::Future`
```

it'd be nice if we could suggest adding `.await` somewhere.

**Comment by nikomatsakis** [comment]
cc @rust-lang/wg-diagnostics @estebank

**Comment by nikomatsakis** [comment]
@wycats was reporting a much worse error, but I'm quite sure what he did to encounter it:

![image](https://user-images.githubusercontent.com/155238/58241249-dcfc0080-7d4c-11e9-80fe-fcb69c052839.png)

Probably has something to do with streams?

**Comment by estebank** [comment]
This will be harder to do for chained expessions `future.field_or_resolved_future.execute_new_future()?` that should have been `future.await.field_or_resolved_future.execute_new_future().await?`.

**Comment by cramertj** [comment]
@nikomatsakis Looks like @wycats's error was the same, it's just that the output type of the future is more complicated (`Pin<Box<Stream<Item = Value> + Send + 'static>>`).

**Comment by doctorn** [comment]
Hi, I'd be really interested in working on this

**Comment by davidtwco** [comment]
@doctorn awesome! Sending `@rustbot claim` should assign you to the issue and feel free to start a topic in the [#wg-async-foundations](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations) stream if you need a hand.

**Comment by gilescope** [comment]
Can we close this now @doctorn ?

**Comment by doctorn** [comment]
Unfortunately not... I haven't had a lot of time for the past few weeks, but basically all cases similar to this one that @estebank pointed out still aren't covered.

`future.field_or_resolved_future.execute_new_future()?` that should have been `future.await.field_or_resolved_future.execute_new_future().await?`.

I've made some progress towards this, but there are some complicated trait predicates involved in solving the `?` case. I brought this up in #t-compiler/help but I'm not aware of anyone who knows how to solve this at the moment.

**Comment by estebank** [comment]
I fear that that case will remain open for a while, as it'll be hard to get right.

**Comment by gilescope** [comment]
I was thinking that this issue was fairly similar to #63100.

**Comment by doctorn** [comment]
It is essentially the same issue. It boils down to altering the trait inference code to give diagnostics when certain necessary trait predicates don’t hold. For example if you consider the question mark case:

```
async fn foo() -> Result<(), ()> {
    Ok(())
}

async fn bar() {
    foo()?
}
```

The error here arises because `Future<Output=Result<(), ()>>` does not implement `std::ops::Try`; however, it’s obvious to us that the `Output` type does, so we want to await here.

What we’d want to do is say that any time type `T` can’t be shown to implement an arbitrary trait (say `Baz`), if we can show that `T: Future<Output=U> where U: Baz` suggest `.await`. Similarly, for functions rather than futures if we can show that `T: Fn() -> U where U: Baz` suggest calling it.

Long story short, we need to verify if a trait bound holds for a projected type; however, I don’t think you can do this easily.

**Comment by doctorn** [comment]
#63100 only points out cases where we have a concrete expected type which are the same cases I covered with #62067. Covering expected trait bounds is where it gets sticky.

**Comment by estebank** [comment]
From #62067, these are the cases still missing:

- [ ] Return places for functions
- [ ] Field access
- [ ] Method invocation

**Comment by nikomatsakis** [comment]
@rustbot release-assignment

@doctorn mentioned they were pretty busy

**Comment by tmandry** [comment]
@csmoe Thanks for improving on this with #71948. Do you want to continue working on this, or release the assignment for now?

**Comment by csmoe** [comment]
@tmandry I'll try to fix them all, more commits pending.

**Comment by csmoe** [comment]
> From #62067, these are the cases still missing:
> 
>     * [ ]  Return places for functions
> 
>     * [ ]  Field access
> 
>     * [ ]  Method invocation
> 
> 
> In bodies that aren't `async`, `.await` is not suggested, but maybe we should mention that we're _not_ in an `async` block and that the future _could_ have been awaited. I guess this is likely to be a common wip error during development.

@estebank could you write a case of `return-place-for-functions`? I'm not sure about this one, since `await` seems already [suggested](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=bcd19deba7041bfbb2bc2ee6c87b032d)

**Comment by estebank** [comment]
Indeed. I'm not sure if it is due to an improvement since august last year, or just a mistake on my part. Either way, we are in a good place today :)

**Comment by tmandry** [comment]
Ping from wg-async-foundations triage. @csmoe are you still working on this?

**Comment by csmoe** [comment]
@tmandry yes, in slow progress since time is tough these days.
